### PR TITLE
KRV-21705: Support on-boarding of brownfield clusters

### PR DESCRIPTION
**Pull request by ashleyvjoy**
When the DN client is deployed through ApexConnectivityClient CR, the csm-operator looks for the existing CSM objects and the operator adds the required roles(similar to the one for dell-csm namespace) for the namespaces where the csm object resides.

This PR addresses identifying the namespace where csm objects are present in the cluster and create the required roles/rolebindings 

https://jira.cec.lab.emc.com/browse/KRV-21705

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |

# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Scenario-1: CSM-Operator version < 1.4(1.2&1.3) was deployed in the cluster and followed by powerflex driver. The operator image was upgraded to the test image and the apex connectivity client was deployed. The namespaces where the csm is installed are identified and the role/rolebindings are created as part of connectivity client deployment

- [x] Scenario-2: The scenario was tested with powerflex and application mobility deployed in the cluster on which the connectivity client was deployed

- [x] Scenario-3: CSM-Operator version 1.4.4 was deployed in the cluster and followed by powerflex driver. The operator image was upgraded to the test image and the connectivity client was deployed. The namespaces where the csm is installed are identified and the role/rolebindings are created as part of connectivity client deployment

- [x] Scenario-4: CSM-Operator version with test image was deployed in the cluster and followed by powerflex driver and the connectivity client . The connectivity client was then re-installed after deleting the role/rolebindings from the cluster. The namespaces where the csm is installed are identified and the role/rolebindings are created as part of connectivity client deployment

1logs12to16a

2logs12to16

### CODE REVIEW

**Comment by ashleyvjoy** [comment]
@jooseppi-luna @donatwork @rajkumar-palani please review these code changes. 
Note: On-boarding of brownfield cluster will be attempted when the call to deploy apex connectivity client is made as part of reconciliation. However, after the cluster is onboarded, there can be a scenario where the user directly installs csi/csm on the cluster. Is that supported? If yes, then call to onboard brownfield cluster(thereby discovering the csm namespace) have to be moved from the DeployApexConnectivityClient() to the main reconciliation function.

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -1105,3 +1110,101 @@ func DetermineUnitTestRun(ctx context.Context) bool {
...
+
+	for _, obj := range deployObjects {
+		log.FromContext(ctx).Info("namespace of parsed object is", "object", obj.GetNamespace())
+
+		if err := ctrlClient.Create(ctx, obj); err != nil {
+			if k8serror.IsAlreadyExists(err) {
+				log.FromContext(ctx).Info("Object already exists", "object", obj.GetObjectKind().GroupVersionKind().String())
+				continue
+			}
+			return fmt.Errorf("failed to create object: %w", err)
```
Comment (donatwork) [inline pkg/utils/utils.go:1205]: What if an object with the name exists but is in a different state than what we need to create? Should an update be done? Also at least log at warning level.

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -1105,3 +1110,101 @@ func DetermineUnitTestRun(ctx context.Context) bool {
 	}
 	return unitTestRun
 }
+
+func BrownfieldDeployment(ctx context.Context, path string, cr csmv1.ApexConnectivityClient, ctrlClient crclient.Client) error {
+	//Get the namespaces
```
Comment (donatwork) [inline pkg/utils/utils.go:1115]: Remove comment.

**Review by donatwork — APPROVED** [review]

File: pkg/utils/utils.go (hunk 1 above)
Comment (ashleyvjoy) [reply pkg/utils/utils.go]: Addressed this

File: pkg/utils/utils.go (hunk 2 above)
Comment (ashleyvjoy) [reply pkg/utils/utils.go]: Done

File: pkg/utils/utils.go [hunk 3]
```diff
@@ -1105,3 +1110,101 @@ func DetermineUnitTestRun(ctx context.Context) bool {
 	}
 	return unitTestRun
 }
+
+func BrownfieldDeployment(ctx context.Context, path string, cr csmv1.ApexConnectivityClient, ctrlClient crclient.Client) error {
```
[the remark this answers is not in the document]
Comment (ashleyvjoy) [reply pkg/utils/utils.go]: Sorry for the confusion. This is just a function to start attempting the brownfield onboarding of clusters. Renamed it to BrownfieldOnboard

**Pull request closed without merging**